### DISCUSSION TITLE
v0.5.0 Beta: Server Side Rendering and Schema Components

### DISCUSSION THREAD

**Discussion by rohit-gohri** [rohit-gohri/redocusaurus#88] [Announcements]
### Minor Changes

- [`0fbefef`](https://github.com/rohit-gohri/redocusaurus/commit/0fbefef62247f1153102254613a120793a329861) Thanks [@rohit-gohri](https://github.com/rohit-gohri)! - Allow importing single model defenition inside mdx files - #46

* [`12d7a26`](https://github.com/rohit-gohri/redocusaurus/commit/12d7a261c2736860fd48d77d5727b0f7127b807b) Thanks [@rohit-gohri](https://github.com/rohit-gohri)! - Add support for SSR - #42

### Patch Changes

- [`e0856fa`](https://github.com/rohit-gohri/redocusaurus/commit/e0856faccaae7ef65768ef7c3453d8a272800562) [#86](https://github.com/rohit-gohri/redocusaurus/pull/86) Thanks [@rohit-gohri](https://github.com/rohit-gohri)! - Fix common types not being exported

<hr /><em>This discussion was created from the release <a href='https://github.com/rohit-gohri/redocusaurus/releases/tag/v0.5.0-next.0'>v0.5.0 Beta: Server Side Rendering and Schema Components</a>.</em>

**Comment by rohit-gohri** [comment]
You will have to use the custom ServerStyle component in your [Theme Root](https://docusaurus.io/docs/using-themes#wrapper-your-site-with-root) (as mentioned in https://github.com/facebook/docusaurus/issues/3236#issuecomment-788953743) for SSR like below:


```tsx
import React from 'react';

import ServerStyle from '@theme/ServerStyle';

function Root({ children }: { children: React.Component }): JSX.Element {
  return (
    <>
	  // TODO: Remove when docusaurus adds proper css-in-js support (https://github.com/rohit-gohri/redocusaurus/issues/90)
      <ServerStyle from={children} />
      {children}
    </>
  );
}

export default Root;
```

See https://github.com/rohit-gohri/redocusaurus/blob/7bf52e9be23da74c1bde920377a081f5efac14a2/website/src/theme/Root/index.tsx for example project.

**Reply by aeneasr** [reply]
Awesome :)

**Reply by rohit-gohri** [reply]
FYI: This is broken in the next beta release of docusaurus AFAIK. So would recommend on staying on current dependencies for a while

**Reply by JohannesFerner** [reply]
> next beta release

Beta.9 or Beta.10? 
Is there currently a way to get SSR running (with ~5 different Specs in my case) running on Beta.9 ?

**Reply by rohit-gohri** [reply]
It was broken on beta.6, everything should be working on beta.9, it's tested and working with the examples: https://github.com/rohit-gohri/redocusaurus/blob/main/website/package.json.

You can read [these docs on how to setup SSR](https://rohit-gohri.github.io/redocusaurus/docs/guides/server-side-rendering)

**Reply by rohit-gohri** [reply]
This is now automatically done in the upcoming [redocusaurus@v1](https://github.com/rohit-gohri/redocusaurus/discussions/146)! (no need for custom `@theme/Root` 🙂). Along with lots of other things.

**Comment by SrMouraSilva** [comment]
Hello, @rohit-gohri. Congratulations for this awesome project.

Is there a configuration to disable SSR? I would like to the browser loads the component when a user access the page.

**Reply by rohit-gohri** [reply]
There is no configuration like that right now. But would be open to adding it if you want to do it

**Reply by rohit-gohri** [reply]
I think someone posted a workaround here - https://github.com/rohit-gohri/redocusaurus/issues/215#issuecomment-1307557779

**Reply by SrMouraSilva** [reply]
Hi @rohit-gohri. Thanks for the quick response.

Maybe I wasn't not clear why I'm trying to do this. In my scenario, I have a REST application that generates the `openapi.json` when it starts based on its endpoints.

I would like to redocusaurus loads its `openapi.json` dynamically when a user loads the page. So, if I update my REST application later and someone access redocusaurus, it will shows the last `openapi.json` version instead of the version that when the documentation was built.

I think that would be possible if there is an option that doesn't process the openapi file on build time.
What do you thing about it?

**Reply by rohit-gohri** [reply]
Ohh.. that makes sense. I think that would already be possible by just using the `@theme/Redoc` component directly and creating your own page. I can add an example for it.

I don't think an option for it makes sense, since features like [schema imports](https://redocusaurus.vercel.app/docs/guides/schema-imports) won't work without build time rendering. And I would still recommend to use build time rendering even in your case, maybe by adding a trigger for the docs ci pipeline to run whenever the REST application does a release.

**Reply by SrMouraSilva** [reply]
> Ohh.. that makes sense. I think that would already be possible by just using the @theme/Redoc component directly and creating your own page. I can add an example for it.

@rohit-gohri, I would appreciate if you provide some example. Actually, I was trying to use, but until now I didn't have any progress.
If you provide some scratch code it would be a great help to me.

**Reply by rohit-gohri** [reply]
I will try to add an example to the docs this week. I think it would require a small change in the code and a release.

**Reply by rohit-gohri** [reply]
The example is live here - https://redocusaurus.vercel.app/examples/client-only/
This is the code - https://github.com/rohit-gohri/redocusaurus/blob/main/website/src/pages/examples/client-only/index.jsx

Only supported in latest v1.5.0 version of the package

**Reply by SrMouraSilva** [reply]
Thanks, @rohit-gohri. It works like a charm.
There is only a thing that isn't clear for me. How can I change redoc options? Eg. I would like to define `expandResponses: "200,201",`

**Reply by rohit-gohri** [reply]
I think that's part of Redocly's paid features (according to their docs - https://redocly.com/docs/api-reference-docs/configuration/functionality/#configuration-options-for-api-docs). We don't support that since we use the OSS package and so are limited to options https://github.com/Redocly/redoc/

If it's mentioned in https://github.com/Redocly/redoc/ props then we can set it otherwise probably not

**Reply by rohit-gohri** [reply]
I was wrong it should work, but still figuring out how

**Reply by rohit-gohri** [reply]
1.6.0 now loads `redocly.yaml` correctly for both plugin and theme so you can just define the options there
